dinput_hook/renderer_utils.cpp

```diff
@@ -0,0 +1,428 @@
+}
+
+renderListShader get_or_compile_renderListShader() {
+    static bool shaderCompiled = false;
+    static renderListShader shader;
+    if (!shaderCompiled) {
+        const char *vertex_shader_source = R"(
+#version 330 core
+
+layout(location = 0) in vec4 position;
+layout(location = 1) in vec4 color;
+layout(location = 2) in vec2 uv;
+
+out vec4 passColor;
+out vec2 passUV;
+
+uniform mat4 projMatrix;
+
+void main() {
+    vec4 posView = position;
+    gl_Position = projMatrix * posView;
+
+    passColor = color;
+    passUV = uv;
+}
+)";
+        const char *fragment_shader_source = R"(
+#version 330 core
+
+in vec4 passColor;
+in vec2 passUV;
+
+out vec4 outColor;
+
+uniform sampler2D tex;
+
+void main() {
+    vec4 texel = texture(tex, passUV);
+    outColor = texel * passColor;
+}
+)";
+        GLuint program = glCreateProgram();
+        GLuint vertex_shader = glCreateShader(GL_VERTEX_SHADER);
+        glShaderSource(vertex_shader, 1, &vertex_shader_source, nullptr);
+        glCompileShader(vertex_shader);
+        GLint status = 0;
+        glGetShaderiv(vertex_shader, GL_COMPILE_STATUS, &status);
+        if (status != GL_TRUE) {
+            int length = 0;
+            glGetShaderiv(vertex_shader, GL_INFO_LOG_LENGTH, &length);
+            std::string error(length, '\0');
+            glGetShaderInfoLog(vertex_shader, error.size(), nullptr, error.data());
+
+            fprintf(hook_log, "%s\n", error.c_str());
+            fflush(hook_log);
+
+            std::abort();
+        }
+
+        GLuint fragment_shader = glCreateShader(GL_FRAGMENT_SHADER);
+        glShaderSource(fragment_shader, 1, &fragment_shader_source, nullptr);
+        glCompileShader(fragment_shader);
+        glGetShaderiv(fragment_shader, GL_COMPILE_STATUS, &status);
+        if (status != GL_TRUE) {
+            int length = 0;
+            glGetShaderiv(fragment_shader, GL_INFO_LOG_LENGTH, &length);
+            std::string error(length, '\0');
+            glGetShaderInfoLog(fragment_shader, error.size(), nullptr, error.data());
+
+            fprintf(hook_log, "%s\n", error.c_str());
+            fflush(hook_log);
+
+            std::abort();
+        }
+
+        glAttachShader(program, vertex_shader);
+        glAttachShader(program, fragment_shader);
+        glLinkProgram(program);
+
+        glGetProgramiv(program, GL_LINK_STATUS, &status);
+        if (status != GL_TRUE)
+            std::abort();
+
+        GLuint VAO;
+        glGenVertexArrays(1, &VAO);
+        GLuint VBO;
+        glGenBuffers(1, &VBO);
+        GLuint EBO;
+        glGenBuffers(1, &EBO);
+
+        shader = {
+            .handle = program,
+            .VAO = VAO,
+            .VBO = VBO,
+            .EBO = EBO,
+            .proj_matrix_pos = glGetUniformLocation(program, "projMatrix"),
+        };
+
+        shaderCompiled = true;
+    }
+
+    return shader;
+}
+
+extern "C" void renderer_drawRenderList(int verticesCount, LPD3DTLVERTEX aVerticies, int indexCount,
```

Comment:
it could be more efficient to update the whole vertex cache in one go every frame. just an idea

Reply:
Ok ? I have no idea how to do that but it would be similar to `render_mesh` buffer creation ?

Reply:
renderer_drawRenderList is called for very small parts, maybe even for every quad (have to check that). but since the game adds all vertices to the global vertex cache, all those could be copied to the gpu in one go in rdCache_SendFaceListToHardware. since you want to replace most of the rendering anyway this is of course not really relevant 🙂